### ISSUE TITLE
CRC32 Algorithm - Explanation

### ISSUE THREAD

**Issue by x1mus** [cfsamson/learning-crc32#1]
Hello !
This issue isn't to point out that something isn't working but since the repository is meant for learning I'll ask my question 😄

I'm currently trying to understand how CRC-32 works if I wanted to calculate it by hand. I first heard of the algorithm digging into PNG so I'm using PNG chunk values for convenience.

I found on the web the steps to follow in order to compute the CRC-32 :
```
- Reverse the input bits --> (1011 becomes 1101)
- Pad the input with degree of polynomial  – 1. --> (No clue what this means)
- XOR with 0xFFFFFFFF (In hardware this is used to set the memory to a known state) --> (Basically inverts the bits. 1011 becomes 0100)
- Start the binary division and stop when you’ve fully divided the divisor into the dividend. --> (Not too sure how to proceed with this)
- Take the remainder and XOR with 0xFFFFFFFF again. --> (Basically inverts the bits 1011 becomes 0100)
- Reverse the bits --> (1011 becomes 1101)
```

Inside the brackets is my understanding of the step so in the process, I'm kinda lost about how to really do it properly.

Here's the value I'm currently working with to make it easier:
```
0x494844520000038C000002D00806000000 --> Input
0xDBF1FE9A --> Expected CRC
```

And I'm working with the polynomial described in the libpng (no surprise here).
`x^32+x^26+x^23+x^22+x^16+x^12+x^11+x^10+x^8+x^7+x^5+x^4+x^2+x+1`

Have a nice day !

**Comment by cfsamson** [comment]
Hi. I have created a CRC-32 version below with a focus on explaining every step of the process. I hope I covered everything you asked for in the comments below.

You can also [test the code in the Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7d22940c57a48d90299ad531309d179f). It's probably easier to read the code there unless you copy it to a text editor yourself and try to run it.

It's a bit different when you do it by hand since you'll not use the truncated polynomial as you do programmatically, but the full polynomial (the truncated removes the initial bit which is always 1). For manually solving the problem I recommend [this video](https://youtu.be/izG7qT0EpBw) as mentioned earlier. For an explanation on how to solve it programmatically read the code below.

`x^32+x^26+x^23+x^22+x^16+x^12+x^11+x^10+x^8+x^7+x^5+x^4+x^2+x+1` should be the same as `0x04C11DB7` but I haven't actually confirmed that.

Let me know if it helps!

```rust
// CRC-32
// Truncated Poly.:     0x04C11DB7
// Initial remainder:   0xFFFFFFFF
// Reflect data:        yes
// Reflect remainder:   yes
// Expected result:     0xDBF1FE9A
//
//        initial reminder                                                                       message
// --------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
// 111111111111111111111111111111110100100101001000010001000101001000000000000000000000001110001100000000000000000000000010110100000000100000000110000000000000000000000000
// 100000100110000010001110110110111
//  |------------------------------|
//    Trunc. polynomial (32 bits)
// |-------------------------------|
//    Full polynomial (33 bits)
//
// - We know that any valid polynomial has to start with a 1 bit so the
//   "interesting" part is everything after the first bit (called the truncated
//   polynomial). See comment in the `crc` function.
fn main() {

    let msg: &[u8] = &[0x49, 0x48, 0x44, 0x52, 0x00, 0x00, 0x03, 0x8C, 0x00, 0x00, 0x02, 0xD0, 0x08, 0x06, 0x00, 0x00, 0x00];
    let crc = crc(0x04C11DB7, 0xFFFFFFFF, 0xFFFFFFFF, msg);

    print!("msg: ");
    for b in msg {
        print!("{:08b}", b);
    }
    print!("\n\n");
    println!("------- Results -------");
    println!("CRC bits: {:032b}", crc);
    println!("CRC hex: {:0X}", crc);
}

/// This is what we do for CRC-32, step by step:
///
/// 1. We make a copy of the remainder that we can change, which will store the
/// final CRC
/// 2. For each byte in the message:
///     1. Reverse the bits og the byte (1100) => (0011)
///     2. We push the message bytes to the leftmost bits of the remainder
///     3. For each of the 8 bits of the "message byte" we:
///         1. Shift the bits 1 position left if the leftmost is a 0
///         2. When the leftmost is a 1, we shift out 1 byte (since it will always end up a 0)
///            then we divide the remainder with the truncated polynomial
/// 3. In CRC-32 we XOR the result with the value 0xFFFFFFF (which is the same as
///    inverting the bits)
/// 4. We reverse the bits of the final remainder to get our CRC
fn crc(poly: u32, remainder: u32, final_xor: u32, message: &[u8]) -> u32 {
    println!("polynomial: {:032b}", poly);
    println!("initial_remainder: {:032b}", remainder);
    println!("final_xor: {:032b}", final_xor);
    println!("remainder: {:032b}", remainder);

    // 1. Just make a mutable variable storing the remainder
    let mut remainder: u32 = remainder;

    // 2. Perform modulo-2 division, a byte at a time
    for b in message {
        let b = *b;
        // 2.1 reverse bits
        let b = b.reverse_bits();

        // 2.2 Shift the reflected message bytes to the leftmost bits of remainder. Ex:
        // beofre: 11111111111111111111111111111111
        // after:  01101101111111111111111111111111
        remainder ^= (b as u32) << (32 - 8);

        // 2.3 For each bit position in the message
        for _ in 0..8 {
            if remainder & 0b10000000000000000000000000000000 == 0b10000000000000000000000000000000 {
                // 2.3.2 (see initial explanation of steps)
                remainder = (remainder << 1) ^ poly as u32;
                //          |--------------|
                //                  ^
                //                  IMPORTANT: because we know that the "full poly"
                //                  actually starts with a 1 (which is not included
                //                  in the trucated poly) it will always "switch" the
                //                  leftmost bit to a 0 whcih is shifted out so we
                //                  actually shift it out here before we divide the
                //                  rest with the truncated poly.
            } else {
                // 2.3.1 Shift the bits 1 step left
                remainder = remainder << 1;
            }
        }
    }

    // 3. XOR with the value 0xFFFFFFFF
    let remainder = remainder ^ final_xor;
    // 4. reverse bits to get the final CRC
    let remainder: u32 = remainder.reverse_bits();
    remainder
}
```

The resulting output is:
```text
polynomial: 00000100110000010001110110110111
initial_remainder: 11111111111111111111111111111111
final_xor: 11111111111111111111111111111111
remainder: 11111111111111111111111111111111
msg: 0100100101001000010001000101001000000000000000000000001110001100000000000000000000000010110100000000100000000110000000000000000000000000

------- Results -------
CRC bits: 11011011111100011111111010011010
CRC hex: DBF1FE9A
```

**Comment by cfsamson** [comment]
As for manually calculating the CRC (besides the video I sent) I think you need to do the following steps (I think):

```
1. Reverse the input bits  for every **byte**--> (10110001 11110000 becomes 10001101 00001111)
2. Pad the input with zeroes in the end with the same length of the CRC [ 0x49..., 0x00, 0x00, 0x00, 0x00]
3. Put the inital remainder before the message so you have [0xFF, 0xFF, 0xFF, 0xFF, 0x49...]
4. Perform modulo 2 division and stop when you’ve fully divided the divisor into the dividend. See explanation below
5. Take the remainder and XOR with 0xFFFFFFFF (Basically inverts the bits 1011 becomes 0100)
6. Reverse the bits (not the bits in the individual bytes) --> ( (10110001 11110000 becomes 00001111 10001101 )
```

These should be the main steps to take, but doing this manually for a 32 bit CRC will take a lot of time and as the process is very error prone. For learning purposes I recommend using a 16 bit CRC or maybe even an 8 or 4 bit CRC so you can do the basic steps but don't have to repeat them a huge number of times. I recommend checking out the manual division [explained here](https://barrgroup.com/Embedded-Systems/How-To/CRC-Calculation-C-Code) for a 4 bit CRC.

Basically, performing step 4 above with the polynomial: `0b11011`, the message and initial remainder of `0b11100101`, you can do the same as we do in code:

```
     msg    padding
   -------|----
   111001010000
  ÷11011
1:=001111010000 # shift left
2:=01111010000  # shift left
  =1111010000   # divide
  ÷11011
3:=0010110000   # shift left
4:=010110000    # shift left
  =10110000     # divide
  ÷11011
5:=01101000    # shift left
  =1101000     # divide
  ÷11011
6:=0000100      # shift left
7:=000100       # shift left
8:=00100        # shift left
  =0100         # stop shifting left when you've shifted left n - padding bits (8 in this case)

CRC = 0b0100
```

**Comment by x1mus** [comment]
Thanks for the explanations, I will definitely try to compute one myself to be sure I'm not stuck at any point :)

**Comment by x1mus** [comment]
Hello !
I tried to implement the method that we would do by hand into a program (like that, I'm less prone to typing errors). So of course, the implementation is very naive and clearly not efficient.

But I'm still having a problem somewhere and I can't figure out why ^^ Here's the code I wrote (python)
I tried to add as much print as possible and some comment for every step, like that you get my thinking process. Thank you in advance even if I know that this may seem easy for you now :x 
```python
def convert_hex_to_binary_list(message):
	# Convert the hex values to a binary string
	binary = "{0:b}".format(int(message, 16))

	# Adding 0 on the left to be sure this a a multiple of 8 bits
	while len(binary) % 8 != 0:
		binary = "0" + binary

	# Return the list of bytes
	return [binary[i:i+8] for i in range(0, len(binary), 8)]


def invert_bits(message):
	inverted_message = ""
	for bit in message:
		if bit == "0":
			inverted_message += "1"
		else:
			inverted_message += "0"

	return inverted_message


def main():
	message = "494844520000038C000002D00806000000"
	poly = "104C11DB7"
	poly_binary = "{0:b}".format(int(poly, 16))
	expected_crc = "DBF1FE9A"
	print("INITIAL VALUES :")
	print("================")
	print("Message in hex :", message)
	print("Poly in hex :", poly)
	print("Poly in binary :", poly_binary)
	print("Expected CRC :", expected_crc)
	print()


	bytes_list = convert_hex_to_binary_list(message)
	print("Initial input message :")
	print("=======================")
	print(bytes_list)
	print()

	"""First step:
	=====================
	- Reverse the input bits for each bytes
	"""
	temp = []
	for byte in bytes_list:
		temp.append(byte[::-1])
	bytes_list = temp
	del temp
	print("Each bytes reversed input message :")
	print("===================================")
	print(bytes_list)
	print()

	"""Second step:
	=====================
	- Adding padding to the input (4 bytes of 0)
	"""
	i = 0
	while i < 4:
		bytes_list.append("00000000")
		i += 1
	print("Padding added to input message :")
	print("================================")
	print(bytes_list)
	print()


	"""Third step:
	=====================
	- XOR with 0xFF (Invert the bits)
	"""
	final_input_binary = ""
	for byte in bytes_list:
		final_input_binary += invert_bits(byte)
	print("Inverting bits to input message :")
	print("===================================")
	print(final_input_binary)
	print()

	"""Fourth step:
	=====================
	- Make the division and save the remainder
	"""
	remainder = bin(int(final_input_binary, 2) % int(poly_binary, 2))[2:]
	print("Computed the remainder :")
	print("========================")
	print(remainder)
	print()

	"""Fifth step:
	=====================
	- XOR the remainder with 0xFF (Invert the bits)
	"""
	remainder = invert_bits(remainder)
	print("Inverted remainder :")
	print("====================")
	print(remainder)
	print()

	"""Last step:
	=====================
	- Reverse the entire binary
	"""
	remainder = remainder[::-1]
	print("Reversed remainder :")
	print("====================")
	print(remainder)
	print()


	"""Displaying results
	"""
	remainder = hex(int(remainder, 2))
	print("Obtained results :")
	print("==================")
	print("Computed CRC :", str(remainder)[2:].upper())
	print("Expected CRC :", expected_crc)
	print()


	print("Are the CRC equivalent ?")
	print("========================")
	if str(remainder)[2:].upper() == expected_crc:
		print("They are !")
	else:
		print("They aren't !")


if __name__ == "__main__":
	main()
```

**Comment by cfsamson** [comment]
Unfortunately I'm not an expert in Python, so take the changes I made for what they are. We're calculating the CRC using string representation of binary here (which I've never done before since it's easier to simply use the CPU instructions for working with the binary data itself).

I think the major cause of confusion is what you called the third step (which I simply renamed):
` XOR with 0xFFFFFFFF (In hardware this is used to set the memory to a known state) --> (Basically inverts the bits. 1011 becomes 0100)`

This is not correct. What the source you got this from is (probably) talking about is the initial state for some CRC algorithms. Usually 32 bit CRC's has a initial value of `0xFFFFFFFF` but that is not the same as inverting all the bits. It's the same as using `11111111111111111111111111111111` for the first division, instead of the polynomial `104C11DB7`. Then performing the rest of the divisions using the polynomial.

Also, the code you wrote for performing modulo 2 division doesn't seem right. % is modulo 10 division if I'm not mistaken? So that won't work. I changed your code so it works here, albeit it's a pretty strange way of doing this it does show a way where you use the computer to do modulo 2 division the same way you would do it by hand (which is good for learning I guess):

```python
def convert_hex_to_binary_list(message):
    # Convert the hex values to a binary string
    binary = "{0:b}".format(int(message, 16))

    # Adding 0 on the left to be sure this a a multiple of 8 bits
    while len(binary) % 8 != 0:
        binary = "0" + binary

    # Return the list of bytes
    return [binary[i:i+8] for i in range(0, len(binary), 8)]


def invert_bits(message):
    inverted_message = ""
    for bit in message:
        if bit == "0":
            inverted_message += "1"
        else:
            inverted_message += "0"

    return inverted_message


def main():
    message = "494844520000038C000002D00806000000"
    poly = "104C11DB7"
    poly_binary = "{0:b}".format(int(poly, 16))
    expected_crc = "DBF1FE9A"
    print("INITIAL VALUES :")
    print("================")
    print("Message in hex :", message)
    print("Poly in hex :", poly)
    print("Poly in binary :", poly_binary)
    print("Expected CRC :", expected_crc)
    print()


    bytes_list = convert_hex_to_binary_list(message)
    print("Initial input message :")
    print("=======================")
    print(bytes_list)
    print()

    """First step:
    =====================
    - Reverse the input bits for each bytes
    """
    temp = []
    for byte in bytes_list:
        temp.append(byte[::-1])
    bytes_list = temp
    del temp
    print("Each bytes reversed input message :")
    print("===================================")
    print(bytes_list)
    print()

    """Second step:
    =====================
    - Adding padding to the input (4 bytes of 0)
    """
    i = 0
    while i < 4:
        bytes_list.append("00000000")
        i += 1
    print("Padding added to input message :")
    print("================================")
    print(bytes_list)
    print()


    """Third step:
    =====================
    - Concatinate the bytes to a string of bits
    """
    final_input_binary = ""
    for byte in bytes_list:
        final_input_binary += byte
    print("Concatinating the bytes to a string of bits :")
    print("===================================")
    print(final_input_binary)
    print()

    """Fourth step:
    =====================
    - Make the division and save the remainder
    """
    print("Modulo 2 division:")
    print("===================================")
    msg_bits = len(final_input_binary) - 32 # remember, we added padding
    # Initial CRC is sometimes used to avoid poor error checking quality in
    # edge cases where the data is simply 0000000...
    initial_crc = "11111111111111111111111111111111"
    remainder = final_input_binary
    counter = 0
    start = True
    while counter < msg_bits:
      if start:
          poly = initial_crc
          start = False
      else:
          poly = poly_binary
      print(remainder)
      print("--------------------------------")
      print(poly)

      interim = ""
      bit_pos = 0
      for bit in remainder:
        if bit_pos < len(poly):
          poly_bit = poly[bit_pos]
          if poly_bit == "1" and bit == "1":
            interim += "0"
          elif poly_bit == "1" or bit == "1":
            interim += "1"
          else:
            interim += "0"
        else:
          interim += bit
        bit_pos += 1
      remainder = interim
      print(remainder)
      # shift zeros out
      first_zeros = 1
      interim = ""
      for bit in remainder:
        if bit == "0" and first_zeros == 1 and counter < msg_bits:
          counter += 1
          print(f"{counter}: shift 1 bit left")
        else:
          interim += bit
          first_zeros = 0

      remainder = interim

    print("Computed the remainder :")
    print("========================")
    print(remainder)
    print()

    """Fifth step:
    =====================
    - XOR the remainder with 0xFF (Invert the bits)
    """
    remainder = invert_bits(remainder)
    print("Inverted remainder :")
    print("====================")
    print(remainder)
    print()

    """Last step:
    =====================
    - Reverse the entire binary
    """
    remainder = remainder[::-1]
    print("Reversed remainder :")
    print("====================")
    print(remainder)
    print()


    """Displaying results
    """
    remainder = hex(int(remainder, 2))
    print("Obtained results :")
    print("==================")
    print("Computed CRC :", str(remainder)[2:].upper())
    print("Expected CRC :", expected_crc)
    print()


    print("Are the CRC equivalent ?")
    print("========================")
    if str(remainder)[2:].upper() == expected_crc:
        print("They are !")
    else:
        print("They aren't !")


if __name__ == "__main__":
    main()
```

**Comment by x1mus** [comment]
Thanks ! Finally understood the whole process !
Took me a while but we manage to do it ahah

**Comment by cfsamson** [comment]
That’s  great 👍.

I think we can leave the issue open in case there is anyone else with the same question. It might help them as well if they stumble upon this issue.
